causallib/time_varying/GFormula.py

```diff
@@ -46,29 +58,108 @@ def estimate_individual_outcome(self,
+        lengths = np.array([len(X), ])
+        pat_id = X['id'][0]
+
+        assert y.shape[1] >= n_obsv, "n_obsv bigger than observed data"
+
+        simulation["ground_truth"] = y.clone()[:, n_obsv - 1: n_obsv - 1 + self.n_steps, ]  # test dataset from n_obsv
+        simulation["pat_id"] = pat_id
+
+        X = X.repeat(self.n_sims, 1, 1)  # N_sims * T * F
+        a = a.repeat(self.n_sims, 1, 1)  # N_sims * T * F
+        lengths = lengths.repeat(self.n_sims)
+
+        t = n_obsv
+        x_t = X[:, :t, :].clone()  # .unsqueeze(1)
+        a_t = a[:, :t, :].clone()  # .unsqueeze(1)
+
+        act_t = treatment_strategy(x_t[:, -1, :], x_t[:, :-1, :], a_t[:, -1, :])
+
+        x_t = T.cat([x_t, a_t], axis=1)
+        last_t = T.cat([x_t[:, -1, :-1], act_t], axis=1)  # bs * F
+        x_t = T.cat([x_t[:, :-1, :], last_t.unsqueeze(1)], axis=1)  # bs * n_obsv * F
+
+        # init all the models
+        self._init_models() #TODO
+
+        # Simulate
+        with T.no_grad():
+            for _idx in range(self.n_steps):
+                t = _idx + n_obsv - 1  # + 1
+
+                out = self._predict(x_t, t)  # TODO
+                # this is X --> becomes prev_X for next round
+                if t < n_margin - 1:
+                    sim_t = out[:, -1, :].unsqueeze(1)
+                    new_t = X[:, (t + 1), :].unsqueeze(1)
+                    act_t = new_t[..., -1].view(-1, 1, 1)  # bs * 1 * 1
+                else:
+                    sim_t = self._apply_noise(out[:, -1, :].unsqueeze(1), t) # bs * 1 * 1
+
+                    # this is A becomes prev_A
+                    prev_act = x_t[:, -1, -1].unsqueeze(1).unsqueeze(1)  # bs * 1 * 1
+                    act_t = treatment_strategy(sim_t[:, -1, 0], sim_t[:, :, 0], a_t[:, -1, 1])
+                    new_t = T.cat([sim_t, prev_act, act_t], axis=-1)  # .unsqueeze(1) # bs * 1 * F
+                x_t = T.cat([x_t, new_t], axis=1)  # bs * (t + 1) * F
+
+                # prediction
+                actual_t = X[:, :(t + 1), :]
+                out = self._predict(actual_t, t)
+                pred_t = out[:, -1, :].unsqueeze(1)
+
+                simulation['actions'].append(act_t)
+                simulation['covariates'].append(sim_t)
+                simulation['prediction'].append(pred_t)
+                simulation['time'].append(t)
+                if t <= n_margin: # and debug:  # == 11:
+                    # print(T.allclose(hidden[0], hidden_p[0]))
+                    # print(T.allclose(hidden[1], hidden_p[1]))
+                    print(T.cat(simulation['prediction'], dim=1).squeeze())
+                    print(T.cat(simulation['covariates'], dim=1).squeeze())
+
+            simulation['actions'] = T.cat(simulation['actions'], dim=1)  # N_sim * n_steps * F-act
+            simulation['prediction'] = T.cat(simulation['prediction'], dim=1)
+            simulation['covariates'] = T.cat(simulation['covariates'], dim=1)  # N_sim * n_steps * F-act
+        # return res
```

Comment:
@ehudkr, I have a confusion here. What should be the return type in this method? As per the framework, the return type is a data frame for predicting only treatments. But here we also have predicted covariates along with treatment values. So could you suggest us how should we move ahead with the return types?